### PR TITLE
feat: Add the `appium:initialDeeplinkUrl` capabilility

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
Based on https://github.com/appium/WebDriverAgent/pull/845

Theoretically this should reduce the duration of session startup for browser sessions as we don't need to perform an extra setUrl API call

### CODE REVIEW

**Comment by mykola-mokhnach** [comment]
@KazuCocoa could you please verify the feature on a real device

**Comment by KazuCocoa** [comment]
Yes, will do

**Comment by KazuCocoa** [comment]
This improved the start-up time, but a new page could be opened with a new tab. This did not occur every time, but I observed this open the URL with a new tab behavior a few times in my testing, although most of times open the given URL with the current opened tab... To be honest, the condition was not clear, but today's master hasn't this open a new tab behavior.

So, probably it would be nice to address the `safariInitialUrl` will open the given URL with the XCTest's openURL for the iOS 16.4+ in the capability's description.

**Review by KazuCocoa — APPROVED** [review]
(this change itself lgtm)

**Comment by KazuCocoa** [comment]
oh, not for this pr itself, but in case if we could detect the given URL was opened with an app (so the active app was not safari), the context might need to be NATIVE_APP since if the session kept the webview context, webview commands such as get the title command hunged up since current opened app did not have webview context.

e.g. open with `"appium:safariInitialUrl": "https://www.youtube.com/watch?v=_agv9_ovmTs"` with browserName safari, then the link opens youtube app. Then, `mobile:activeAppInfo` was `com.google.ios.youtube` after the new session request, but then, the current context kept WebView. The available contexts did not include WebView, so get a title tried to send the command to remote debugger in XCUITest driver but it did not get any response properly by the remote debugger. Then no response behavior occurred.

**Review by KazuCocoa — CHANGES_REQUESTED** [review]
oh, sorry. I tested this further, then noticed this method opened a URL with Safari. So even the `safariInitialUrl` was `https://www.youtube.com/watch?v=_agv9_ovmTs`, for example, the link did not open youtube app while current master opened the app. So this change breaks existing deeplink behavior with `safariInitialUrl`

**Comment by mykola-mokhnach** [comment]
> then, master branch opened the page with youtube app, but this pr opened the URL in Safari.

To me it looks like it sounds logical if Safari is opened as we have requested a test with Safari, not the youtube app. Means the current master behaviour is not what we want. E.g if one wants youtube app they could always provide the corresponding bundle Id in capabilities.

The problem is the behaviour would still be different for different iOS versions with and without deeplinking support :(

I could probably put it under a new flag capability for now to keep the backward compatibility.
Thanks for checking it.

**Comment by mykola-mokhnach** [comment]
Changed the behaviour, so it is only different if `appium:initialDeeplinkUrl` capability value is provided

File: docs/reference/capabilities.md [hunk 1]
```diff
@@ -27,6 +27,7 @@ about capabilities, refer to the [Appium documentation](https://appium.io/docs/e
 | <div style="width:10em">Capability</div> | Description |
 | --- | --- |
 | `appium:bundleId` | Bundle identifier of the app under test, for example `com.mycompany.myapp`. The capability value is calculated automatically if `app` is provided. If neither `app` or `bundleId` capability is provided then XCUITest driver starts from the Home screen. |
+| `appium:initialDeeplinkUrl` | A deeplink URL used to run either the application assigned to `appium:bundleId`, or the default application assigned to handle the particular deeplink protocol if `appium:bundleId` is not set. If provided in combination with `browserName=safari` then makes Safari to start with the given URL preloaded, which speeds up the session startup. The value of `appium:initialSafariUrl` capability is ignored in such case. An error is thrown on session init if either the value of the capability is not a valid URL, or XCTest was not able to associate it with any existing app, or the actual iOS version is below *16.4* |
```
Comment (KazuCocoa) [inline docs/reference/capabilities.md:30]: It makes sense.

Btw, with youtube as bundleId app and `initialDeeplinkUrl` did not open the given URL (just default youtube home screen) with ios 17, so the behavior could depend on OS versions as well, actually. No bundleId, and only `initialDeeplinkUrl` worked pretty well. So I wondered if we could address no-`appium:bundleId` as primary usage with this... (I mean dropping _the application assigned to appium:bundleId_ in the description)


With `browserName: safari` opened the app with `initialDeeplinkUrl` as same as `safariInitialUrl` this morning, so really not sure what affected the behavior, but safariInitialUrl/initialDeeplinkUrl could behave the same... Then, the user should switch the context properly to NATIVE_APP to interact with opened app with the URL.

So, I wondered if we could improve this documentation with:

- remove "the application assigned to `appium:bundleId`"
    - to reduce possible confusion what I mentioned the above's behavior diff
- add to suggest adjusting the `context` to NATIVE_APP for `initialDeeplinkUrl` and (`safariInitialUrl` as well..?) for the deeplink behavior with browserName=safari

**Review by KazuCocoa — APPROVED** [review]
otherwise lgtm

File: docs/reference/capabilities.md (hunk 1 above)
Comment (mykola-mokhnach) [reply docs/reference/capabilities.md:30]: I try to not do preliminary optimizations and just observe how people are going to use it. We may always improve that afterwards based on the actual feedback.

**Pull request merged**